### ISSUE TITLE
Workers disconnection handling

### ISSUE THREAD

**Issue by alexeygolev** [prdn/pigato#9]
It very well might be that I'm not understanding the source code correctly. If so just point me in the right direction (source/tests).
Let's say I have too workers that can handle the same service prefix messages. I assume if one goes down then a message from a client should be routed to another one. As far as I understand we're doing heartbeating from router to dealer. So after 3 failed responses we just pop workers identity from our workers list. Do I understand correctly that until we remove this worker from our list it is still the preferred worker? (it means that we first need to wait until router marks it as failed and only then we'll be able to make another request from client). Or am I missing something?
Thank you.
PS. really impressed with this project... was about to write my own abstraction for zmq bindings but then found this gem:)

**Comment by prdn** [comment]
Hello,

First of all thank you for your interest.
Requests are routed to free workers (for a service) with an LRU policy.
It means that the requests should be equally distributed across workers.

However if a worker dies the request that it is running on that worker
doesn't reply to the respective client and a timeout is sent in its place.
We can handle this behaviour letting a client to specify an option called
'persist' or 'diehard' :-) that tells to the broker to route a request to
another worker if the previous one dies.

However I think that that flag should be false by default. Think about a
worker that takes a request and insert a row in a database. The worker may
die before replying to the client but after inserting the row in the db. So
it's safer to keep the flag off and let developers to choose the right
behaviour.

**Comment by alexeygolev** [comment]
I agree on the db scenario. Was just trying to understand the inner workings. I tried to do a 'retry' from the client when he receives a timeout error, however the first thing the client receives is and 'undefined' response. So the scenario is:
  - Spin up one client, broker and two workers
  - Send a request to the broker (reply will be received)
  - Take one worker down
  - Send another request (before the broker removes the reference to a dead worker — so by default in 7500 timeframe)
 - When C_TIMEOUT received by the client send the same request (I tried both — closing the socket and sending the req with a new one and just resending the req without closing)
- The dealer will receive two messages... on with undefined payload and the next one with a proper reply.

This ghost message I would assume is the message sent when the worker is dead. And taken it makes its way to the dealer it has a proper envelope. So I guess the req is being kept in the queue and then sent back on the next reply. Or there is always an option that I'm doing something wrong and not understanding the inner workings.

**Comment by prdn** [comment]
When  you day 'the dealer receives two messages' you mean the client right?
In that case is weird. I never experienced that issue.
Tomorrow I'll test myself trying to reproduce the bug.
Could you test again running all the scripts with debug activated?

example
$ DEBUG="PIGATO:*" node broker

Do you notice any strange behaviour?

**Comment by alexeygolev** [comment]
Yes, by 'dealer' I mean client (was thinking about the actual architecture when writing, sorry). Tried running with debug on. Nothing suspicious. Will do more tests today.

**Comment by alexeygolev** [comment]
Well I couldn't prove anything with tests... I guess it's in my code. Will keep checking.
Will fire up a pull request for a new test today (just so we have a proof for a scenario like mine)

**Comment by prdn** [comment]
Sounds good to me.
I'm going to push a couple of fixes and the persist flag implementation today or tomorrow.

**Comment by alexeygolev** [comment]
Good, thanks. Added a test. Closing this for now, will reopen if need be. Off to sort out my code:)

**Comment by alexeygolev** [comment]
Figured it out. When using streams instead of callbacks 'error' event doesn't close the stream. Null is pushed to the stream so 'end' event is fired. I was accumulating data in 'data' event and as 'data' event never fired data was undefined. Is it intentional? I guess we can't avoid the 'end' event when using streams. Here is a test that fails:
```js
 it('Fires end event after error only after new request', function (done) {
    var ns = uuid.generate();
    this.timeout(15000);
    var client = new PIGATO.Client(location);
    var worker = new PIGATO.Worker(location, ns);
    var backupWorker = new PIGATO.Worker(location, ns);
    var chunk = 'foo';
    client.start();
    worker.start();

    worker.on('request', function(inp, res) {
      res.end(inp + ':bar');
    });

    backupWorker.on('request', function(inp, res) {
      res.end(inp + ':bar');
    });

    var counter = 0;
    function request() {
      var dataResult;
      client.request(ns, chunk, {timeout: 5000})
        .on('data', function (data) {
          chai.assert.equal(data, chunk + ':bar');
          dataResult = data;
          chai.assert.equal(dataResult, chunk + ':bar');
        })
        .on('end', function () {
          if(counter >= 3) {
            stop(null);
          } else {
            setTimeout(function () {
              worker.stop();
              request();
            }, 2000)
          }

          chai.assert.equal(dataResult, chunk + ':bar');
          counter++;
        })
        .on('error', function (err) {
          if (err === 'C_TIMEOUT' && counter < 3) {
            request();
          } else {
            stop('All workers died');
          }
        });
    }

    request();
    setTimeout(function () {
      backupWorker.start();
    }, 10000);

    function stop(err) {
      client.stop();
      worker.stop();
      done(err);
    }

  });
```
For some reason though the 'end' event for the 'error' is being fired only when the next request is sent by the client:
  - Client sends a request
  - Receives nothing
  - Timeout error
  - Resend the request
  - 'end' event fired (without firing 'data' so I'm pretty sure it's the 'end' event from error)
  - 'data' event fired (taken worker is online) and another 'end' event fired

**Comment by prdn** [comment]
Using console.logs I have traced this behaviour:
- REQ#1  (elapsed 0ms)
- END REQ#1 (elapsed 3ms)
- STOP WRK#1 (elapse 2005ms)
- REQ#2 (elapsed 2005ms)
- ERR TIMEOUT REQ#2 (elapsed 7514ms)
- END REQ#2 (elapsed 7514ms) 

In this last event (END REQ#2) 'dataResult' is undefined (and it is correct because the timeout).

As stated here http://www.bennadel.com/blog/2678-error-events-don-t-inherently-stop-streams-in-node-js.htm an error in a Stream doesn't imply the end of the stream. So the end is called when the stream is effectively ended.

I have added a couple of console logs to your code.
```
   it('Fires end event after error only after new request', function (done) {
    var ns = uuid.generate();
    this.timeout(20000);
    var client = new PIGATO.Client(location);

    var worker = new PIGATO.Worker(location, ns);
    var backupWorker = new PIGATO.Worker(location, ns);
    
    var chunk = 'foo';
    client.start();
    worker.start();

    worker.on('request', function(inp, res) {
      res.end(inp + ':bar');
    });

    backupWorker.on('request', function(inp, res) {
      res.end(inp + ':bar');
    });

    var d1 = new Date();

    var counter = 0;
    function request() {
      var dataResult;
      client.request(ns, chunk, { timeout: 5000 })
        .on('data', function (data) {
          chai.assert.equal(data, chunk + ':bar', "fail1");
          dataResult = data;
          chai.assert.equal(dataResult, chunk + ':bar', "fail2");
        })
        .on('end', function () {
          if(counter >= 3) {
            stop(null);
          } else {
            setTimeout(function () {
              console.log("FIRING REQ", (new Date() - d1));
              worker.stop();
              request();
            }, 2000)
          }
          console.log("END", (new Date() - d1), dataResult);

          chai.assert.equal(dataResult, chunk + ':bar', "fail3");
          counter++;
        })
        .on('error', function (err) {
          console.log("ERR", err);
          if (err === 'C_TIMEOUT' && counter < 3) {
            console.log("FIRING REQ (ERR)", (new Date() - d1));
            request();
          } else {
            stop('All workers died');
          }
        });
    }

    request();
    setTimeout(function () {
      backupWorker.start();
    }, 10000);

    function stop(err) {
      client.stop();
      worker.stop();
      done(err);
    }

  });
```

**Comment by alexeygolev** [comment]
I totally get the 'end' after 'error' behaviour being inherent to streams. What I don't get is this

```
FIRING REQ#1  0
END REQ#1  39 foo:bar
FIRING REQ#2  2042
ERR C_TIMEOUT
FIRING REQ#3  (ERR) 7509
END REQ#2  7509 undefined
```
We're getting 'end' event **after** we're sending the third request (inside 'err' event callback). Even if I add a timeout for the request inside 'error' callback the 'end' event is fired at exactly the same time I'm firing this request. Sorry for asking so many questions.

**Comment by prdn** [comment]
The 'error' event is triggered before the 'end' event because as previously
said an 'error' doesn't destroys a stream.
A normal workflow is:
"data, data, data, end"
An error workflow is:
"data, data, error, data, end" in case of a stream read error
or
"error, end" in the case of a timeout.
So, you've got your error callback called and there you instantly fire a
new request and after that your 'end' callback will be fired.

It seems correct to me.

No problem.

**Comment by alexeygolev** [comment]
So the 'end' callback is kind of 'waiting' for me to fire a new request?

**Comment by prdn** [comment]
You can send your requests whenever you want, but in your case it is safer
send requests in the 'end' event.
Have you seen the new 'retry' option and the its test case?

It could be helpful for you

**Comment by alexeygolev** [comment]
Just checked it out! Amazing, thank you. I guess I should learn more about 'how streams work'. It will probably help me with my confusion. Anyway, closing this. I'm going to clean up my test soon to conform with your style (spawning workers with a helper function etc). Will create a new pull request when done.

**Comment by prdn** [comment]
Cool thank you.
